File: algorithms/0290.word-pattern/solution.rb
# @param {String} pattern
# @param {String} s
# @return {Boolean}
def word_pattern(pattern, s)
  hash = {}
  words = s.split
  chars = pattern.chars

  return false unless words.size == chars.size

  chars.each_with_index do |c, i|
    if hash[c]
      return false unless hash[c] == words[i]
    else
      hash[c] = words[i]
    end
  end

  values = hash.values.sort
  values.size == values.uniq.size
end

File: algorithms/0290.word-pattern/solution.ts
function wordPattern(pattern: string, s: string): boolean {
  const chars: string[] = pattern.split("");
  const words: string[] = s.split(" ");

  if (words.length != chars.length) {
    return false;
  }

  let map = new Map<string, string>();
  for (let i = 0; i < chars.length; i++) {
    const c = chars[i];
    const word = words[i];

    if (map.has(c)) {
      if (map.get(c) != word) {
        return false;
      }
    } else {
      map.set(c, word);
    }
  }

  const values = Array.from(map.values());
  return values.length == new Set([...values]).size;
}
